### ISSUE TITLE
return impl Trait should only permit unique lifetimes

### ISSUE THREAD

**Issue by DutchGhost** [rust-lang/rust#54330]
Providing more than 1 exactly the same lifetime to a return impl Trait, does not give a warning, nor does it give a compiler error:

```Rust
fn example() -> impl Send + Send + 'static + 'static { }

fn main() {
    example();
}
```

The box version does give a warning however:
```Rust
fn dyn_example() -> Box<dyn Send + Send + 'static + 'static> {
    Box::new(())
}

fn main() {
    dyn_example();
}
```
```
 Compiling playground v0.0.1 (/playground)
error[E0226]: only a single explicit lifetime bound is permitted
 --> src/main.rs:1:53
  |
1 | fn dyn_example() -> Box<dyn Send + Send + 'static + 'static> {
  |                                                     ^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0226`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

**Comment by Centril** [comment]
Something is wrong here, but it seems to me that it is:

```rust
fn dyn_example() -> Box<dyn Send + 'static + 'static> {
    Box::new(())
}
```

which is wrong.

Ostensibly, the lifetime derived should be the longest possible of the specified lifetimes, i.e. `'a + 'b = 'a` iff `'a: 'b`.

The consistent thing to do seems to me to warn if duplicate lifetimes are mentioned, or possibly warn iff for any pair of lifetimes `'a` and `'b` mentioned `'a: 'b` holds.

**Comment by rkruppe** [comment]
1. I do not think this is substantially different from #54328 -- we simply do not ever check bounds for redundant entries.

2. (This question goes for the other issue as well:) *Why* should this be an error or warning? It's pointless, but will calling it out prevent any errors?

**Comment by estebank** [comment]
Furthermore, even if we added a lint/warning/error about this we would have to disable them for macro contexts.
